src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala

```diff
@@ -582,10 +582,12 @@ trait TypeDiagnostics extends splain.SplainDiagnostics {
               case bind @ Bind(`n`, _) => bind.updateAttachment(NoWarnAttachment)
               case _ =>
             }
+          def allow(n: Name): Unit = cases.foreach(k => allowVariableBindings(n, k.pat))
           def loop(selector: Tree): Unit =
             selector match {
-              case Ident(n) => cases.foreach(k => allowVariableBindings(n, k.pat))
+              case Ident(n) => allow(n)
               case Typed(expr, _) => loop(expr)
+              case Select(This(_), n) => allow(n)
```

Comment:
Should it drill thru any selection, not just `this.x`?

Reply:
The thinking here is that it's referred as a simple name, so the new name is shadowing.